test/Microsoft.ML.Functional.Tests/DataIO.cs

```diff
@@ -0,0 +1,353 @@
+
+            // Create a lookup table for the types and counts of all properties
+            var types = new Dictionary<string, Type>();
+            var counts = new Dictionary<string, int>();
+            foreach (var property in toyClassProperties)
+            {
+                if (!property.PropertyType.IsArray)
+                    types[property.Name] = property.PropertyType;
+                else
+                {
+                    // Construct a VBuffer type for the array
+                    var vBufferType = typeof(VBuffer<>);
+                    Type[] typeArgs = { property.PropertyType.GetElementType() };
+                    Activator.CreateInstance(property.PropertyType.GetElementType());
+                    types[property.Name] = vBufferType.MakeGenericType(typeArgs);
+                }
+                    
+                counts[property.Name] = 0;
+            }
+
+            foreach (var column in toyDataset.Schema)
+            {
+                Assert.True(types.ContainsKey(column.Name));
+                Assert.Equal(1, ++counts[column.Name]);
+                Assert.Equal(types[column.Name], column.Type.RawType);
+            }
+
+            // Make sure we didn't miss any columns
+            foreach (var value in counts.Values)
+                Assert.Equal(1, value);
+        }
+
+        private IEnumerable<ToyDataset> GenerateToyDataset(int numExamples = 5, int seed = 1)
+        {
+            var rng = new Random(seed);
+            for (int i = 0; i < numExamples; i++)
+            {
+                yield return new ToyDataset
+                {
+                    Label = rng.NextDouble() > 0.5,
+                    Features = new float[] {
+                        (float)rng.NextDouble(),
+                        (float)rng.NextDouble(),
+                        (float)rng.NextDouble(),
+                        (float)rng.NextDouble(),
+                        (float)rng.NextDouble()
+                    },
+                    I1 = (sbyte)rng.Next(),
+                    U1 = (byte)rng.Next(),
+                    I2 = (short)rng.Next(),
+                    U2 = (ushort)rng.Next(),
+                    I4 = rng.Next(),
+                    U4 = (uint)rng.Next(),
+                    I8 = (long)rng.Next(),
+                    U8 = (ulong)rng.Next(),
+                    R4 = (float)rng.NextDouble(),
+                    R8 = (double)rng.NextDouble(),
+                    Tx = GetRandomRomChar(rng),
+                    Ts = TimeSpan.FromSeconds(rng.NextDouble() * (1+rng.Next())),
+                    Dt = DateTime.FromOADate(rng.Next(657435, 2958465)),
+                    Dz = DateTimeOffset.FromUnixTimeSeconds((long)(rng.NextDouble() * (1 + rng.Next()))),
+                    Ug = new RowId((ulong)rng.Next(), (ulong)rng.Next())
+                };
+            }
+        }
+
+        private ReadOnlyMemory<char> GetRandomRomChar(Random rng, int length = 10)
```

Comment:
>GetRandomRomChar [](start = 37, length = 16)

func name unclr - expand ROM

Reply:
Used the `Span` naming convention from .NET.

---
In reply to: [256211211](https://github.com/dotnet/machinelearning/pull/2518#discussion_r256211211) [](ancestors = 256211211)